### PR TITLE
Add more unit tests for socket

### PR DESCRIPTION
**Pull request by wggihan**

### CODE REVIEW

File: stdlib/io/src/main/java/org/ballerinalang/stdlib/io/socket/client/BindAddress.java [hunk 1]
```diff
@@ -69,7 +69,7 @@ public void execute(Context context) {
         } catch (IOException e) {
             String message = "Error occurred while bind the socket address: " + e.getMessage();
```
Comment (hasithaa) [inline stdlib/io/src/main/java/org/ballerinalang/stdlib/io/socket/client/BindAddress.java]: We need to avoid using e.getMessage(), because this can cause the error message to have java traces.  IOException is a supertype, the possibility of having java traces is high.

File: tests/ballerina-unit-test/src/test/resources/logging.properties [hunk 2]
```diff
@@ -22,7 +22,7 @@ java.util.logging.ConsoleHandler.level=ALL
 java.util.logging.ConsoleHandler.formatter=org.ballerinalang.logging.formatters.BallerinaLogFormatter
 org.ballerinalang.logging.formatters.BallerinaLogFormatter.format=%1$tY-%1$tm-%1$td %1$tH:%1$tM:%1$tS,%1$tL %2$s [%3$s] - %4$s %n
 
-org.ballerinalang.test.utils.TestLogHandler.level=INFO
+org.ballerinalang.test.utils.TestLogHandler.level=FINER
```
Comment (hasithaa) [inline tests/ballerina-unit-test/src/test/resources/logging.properties]: This cause to print all Debug logs while building the unit tests. Are we testing any debug log specific test case ? Please revert this change.

File: stdlib/io/src/main/java/org/ballerinalang/stdlib/io/socket/client/BindAddress.java (hunk 1 above)
Comment (wggihan) [reply stdlib/io/src/main/java/org/ballerinalang/stdlib/io/socket/client/BindAddress.java]: Since we are dealing with the outbound network, we need every bit of information in an error situation. Simple we can't use anything and that will make debugging harder when it comes to the narrow down the issue. What is the alternative that we need to use here?

File: tests/ballerina-unit-test/src/test/resources/logging.properties (hunk 2 above)
Comment (wggihan) [reply tests/ballerina-unit-test/src/test/resources/logging.properties]: Yes, this is purposefully added. Debug available only for the IO package. The rest will have their default behavior.

Comment (hasithaa) [reply tests/ballerina-unit-test/src/test/resources/logging.properties]: what is the purpose of having debug log for IO package?

Comment (wggihan) [reply tests/ballerina-unit-test/src/test/resources/logging.properties]: There are some meta level information maintain in the channels and sockets. These things will not interest to the end user and they don't use it. But native developers need this information to verify some information. Also make sure to debug logs give necessary information. We don't have any other mechanism to test whether debug working unless we test that manually.

File: stdlib/io/src/main/java/org/ballerinalang/stdlib/io/socket/client/BindAddress.java (hunk 1 above)
Comment (hasithaa) [reply stdlib/io/src/main/java/org/ballerinalang/stdlib/io/socket/client/BindAddress.java]: You are doing log.error(message, e); after this line. That can be used for troubleshooting. What I am saying Ballerina error message can contain java traces, Because IOException is a generic exception in Java. instead of just using e.getMessage(), we need a proper logic to give a better error message to Ballerina user.

I see this pattern has been used in many native stdlib functions. We need to fix this properly. Please create a separate task for this. 

https://docs.oracle.com/javase/7/docs/api/java/io/IOException.html

File: tests/ballerina-unit-test/src/test/resources/logging.properties (hunk 2 above)
Comment (hasithaa) [reply tests/ballerina-unit-test/src/test/resources/logging.properties]: Ok, I have asked a wrong question. You can have debug logs in your package. I meant to ask was, what is the purpose of enabling debug logs only for IO pkg in test cases. ? 

Manually testing is not a practical and enabling debug logs in unit tests not going to test it either. Write a proper test case then.

File: stdlib/io/src/main/java/org/ballerinalang/stdlib/io/socket/client/BindAddress.java (hunk 1 above)
Comment (wggihan) [reply stdlib/io/src/main/java/org/ballerinalang/stdlib/io/socket/client/BindAddress.java]: That's the main issue. Since IOException is a generic one, we can't pinpointly say this is the issue beforehand. So ballerina developer might be clueless regarding the actual error. So what is the our practice? Give generic message through ballerina error and print stack trace in the ballerina-internal.log?

File: tests/ballerina-unit-test/src/test/resources/logging.properties (hunk 2 above)
Comment (wggihan) [reply tests/ballerina-unit-test/src/test/resources/logging.properties]: I asked [1] the current following practices when it comes to log in debug mode when it come to ballerina native development. Based on an outcome of that discussion, we can either remove or keep the debug log. 

Enabling debug in the test will give us guarantees of the execution of the log without any issue and we can further improve the test by verifying the log whether we are getting expected outcome.

[1] - https://ballerina-platform.slack.com/archives/C9WMA6QDC/p1535002330000100

File: stdlib/io/src/main/java/org/ballerinalang/stdlib/io/socket/client/BindAddress.java (hunk 1 above)
Comment (hasithaa) [reply stdlib/io/src/main/java/org/ballerinalang/stdlib/io/socket/client/BindAddress.java]: I found another bug. If you, look at this particular code, "socket.bind" causes to throw IOExcpetion. If you look at bind methods documentation, you can see there are several checked and uncheck exceptions thrown. In this logic, we have handled only IOExceptions, But there are several runtime errors, none of them handled here. 

Coming back to your original question: the developer should be aware of what kind of errors that can be expected in his logic.  In the above example, if he read the bind documentation he could see what are the possible errors he can get and handle them properly.

File: tests/ballerina-unit-test/src/test/resources/logging.properties (hunk 2 above)
Comment (hasithaa) [reply tests/ballerina-unit-test/src/test/resources/logging.properties]: Ok, I understood your intention now. you want to improve the code coverage by enabling debug logs. But will it test the functionality?

File: stdlib/io/src/main/java/org/ballerinalang/stdlib/io/socket/client/BindAddress.java (hunk 1 above)
Comment (wggihan) [reply stdlib/io/src/main/java/org/ballerinalang/stdlib/io/socket/client/BindAddress.java]: We can simply handle that by catching the Throwable like we do in other places. But still the original question remains that the message can expose other stuff. So my question again "Give generic message through ballerina error and ask developer to check in the ballerina-internal.log?"

Ture. The developer should aware what kind of errors come from the code. This can be more suitable specially core development, like we do develop all the stuff without much depending on the native. But IO is the other handle. We totally depend on the functionalities that provided by native and even we handle the exception separately, for the IOException there can be many.

File: tests/ballerina-unit-test/src/test/resources/logging.properties (hunk 2 above)
Comment (hasithaa) [reply tests/ballerina-unit-test/src/test/resources/logging.properties]: I still don't buy your point. If we enable debug and that causes NPE means bad coding. The developer has to check for possible null pointers, even it is a tracing code.

Still, you don't need to change global log level. Setting log level for your classes will be sufficient to invoke hit your code.

File: stdlib/io/src/main/java/org/ballerinalang/stdlib/io/socket/client/BindAddress.java (hunk 1 above)
Comment (hasithaa) [reply stdlib/io/src/main/java/org/ballerinalang/stdlib/io/socket/client/BindAddress.java]: If other developers can do why you can't do that? https://github.com/ballerina-platform/ballerina-lang/pull/10120/files#diff-3d428a91487a0dfe643619a3bb2722b5L39

I am not saying to give a generic error message.  See above PR. @shafreenAnfar @afkham @sameerajayasoma

Comment (hasithaa) [reply stdlib/io/src/main/java/org/ballerinalang/stdlib/io/socket/client/BindAddress.java]: We don't encorrage people check for "ballerina-internal.log". Given error message should have nessary information for user to understand what went wrong.

File: tests/ballerina-unit-test/src/test/resources/logging.properties (hunk 2 above)
Comment (wggihan) [reply tests/ballerina-unit-test/src/test/resources/logging.properties]: When the developer does the initial implementation, developer know where it can cause the issues. So there can be certain areas that aren't giving any issues at all that time. They can use this information for debug purpose as well. But later due to changes in some other areas this can affect. So based on your logic, we have to do the null check for every variable in every line which is not scalable. Which exactly happen in the past as well.

This will not change the global log level. Apart from this you need to add the package again with the necessary log level. Just simply changing this won't active debug globally.

Comment (hasithaa) [reply tests/ballerina-unit-test/src/test/resources/logging.properties]: Why checking null is not scalable? This is a serious statement. Please schedule a code review for Sockets implementation next week.

File: stdlib/io/src/main/java/org/ballerinalang/stdlib/io/socket/client/BindAddress.java (hunk 1 above)
Comment (wggihan) [reply stdlib/io/src/main/java/org/ballerinalang/stdlib/io/socket/client/BindAddress.java]: Sorry I'm confusing. In above mention PR also has code like this "throw new BLangRuntimeException("ballerina: error reading program file: '" + e.getMessage() + "'", e);".  That's not propagate any java specific code as you mention above?

Comment (hasithaa) [reply stdlib/io/src/main/java/org/ballerinalang/stdlib/io/socket/client/BindAddress.java]: Are you reading the code properly? Before going into 2nd catch, read the first catch block. It is handling two sperate IOExpcetions before handling IOexceptions generally.

Comment (wggihan) [reply stdlib/io/src/main/java/org/ballerinalang/stdlib/io/socket/client/BindAddress.java]: But your argument is not going to use e.getMessage at all right? Also, this is the file reading. It handles the basic stuff like FileNotFound. But there can be various exceptions like content corrupted etc. So there is a high chance that you will hit the 2nd catch. I can add the similar thing to handle most of the stuff, but that's not preventing using the e.getMessage at the end because we don't know the every bit of details that native code does.

Comment (hasithaa) [reply stdlib/io/src/main/java/org/ballerinalang/stdlib/io/socket/client/BindAddress.java]: Do you agree your code has a bug or not? Have you handled all the errors in your code? Please fix that first. Then talk about my first argument.

Comment (wggihan) [reply stdlib/io/src/main/java/org/ballerinalang/stdlib/io/socket/client/BindAddress.java]: I need a proper answer to your first argument, in order to handle this properly. Otherwise, we will end up in a same situation with a couple of more clauses. If our answer to check the ballerina-internal.log and not using the e.getMessage() then I can simply add a meaningful generic message and log the error. But the user might want to check the log to get more information in some situation.

Comment (hasithaa) [reply stdlib/io/src/main/java/org/ballerinalang/stdlib/io/socket/client/BindAddress.java]: I can't repeat the same thing again.  Read this comment. https://github.com/ballerina-platform/ballerina-lang/pull/10157#discussion_r212197848

Comment (wggihan) [reply stdlib/io/src/main/java/org/ballerinalang/stdlib/io/socket/client/BindAddress.java]: I think we are going circle here. First, you mention not to use e.getMessage(). Then you provide a PR which simply does a small check to do a specific message and again propagate rest of the thing that eventually use the e.getMessage(). You are also telling not advisable to use ballerina-internal.log as well. I acknowledge that this error handling can be improve more. But based on current circumstance, we can't avoid the usage of e.getMessage(). If you see any issue, then please provide an alternative.

Comment (hasithaa) [reply stdlib/io/src/main/java/org/ballerinalang/stdlib/io/socket/client/BindAddress.java]: Please schedule a meeting and code review next week.

Comment (wggihan) [reply stdlib/io/src/main/java/org/ballerinalang/stdlib/io/socket/client/BindAddress.java]: Ok Since you also don't have the alternative approach, let's schedule a meeting and come to a final conclusion. My main requirement is not losing any exception information that comes from the native layer.

Comment (shafreenAnfar) [reply stdlib/io/src/main/java/org/ballerinalang/stdlib/io/socket/client/BindAddress.java]: Guys, correct me If I am wrong :) 

We have a rule to make sure not to leak any Javaness to Ballerina user (the programmer who writs Ballerina programs). When we simply say e.getMessage(), there is a chance that we leak Javaness to Ballerina user. Hence, we need to be mindful of this matter and address it as needed. 

At the same time, as Ballerina language developers, we maybe interested in more additional details of the error in order to make debug process easier (consider a production L1). Therefore, in such cases, we can log any required detail in ballerina-internal.log. Even if it contains any Javaness. Because ballerina-internal.log is primarily for Ballerina language developers and not for Ballerina users.
